app/services/pii/encryptor.rb

```diff
@@ -0,0 +1,115 @@
+module Pii
+  class Encryptor
+    DELIMITER = '.'.freeze
+    DIGEST = OpenSSL::Digest::SHA256.new.freeze
+    PADDING = OpenSSL::PKey::RSA::PKCS1_OAEP_PADDING
+
+    # structure of the encrypted payload:
+    #
+    #  cek_encrypted_with_server_key + user_payload_encrypted_with_server_key
+    #                                  v
+    #          (user_private_key_pem + user_ciphertext) + user_payload_signature
+    #                                  v
+    #    cek_encrypted_with_user_key + pii_payload_encrypted_with_user_key
+    #                                  v
+    #                    PII_plaintext + PII_plaintext_signature
+    #
+    # The initial plaintext PII is encrypted once with a user-password-encrypted
+    # private key, then again with the server-password-encrypted private key.
+    # The user private key is encrypted inside the server-encrypted payload.
+
+    def initialize
+      @key_maker = Pii::KeyMaker.new
+      @cipher = OpenSSL::Cipher.new 'AES-256-CBC'
+    end
+
+    def encrypt(plaintext, password)
+      user_private_key_pem = key_maker.generate(password)
+      user_private_key = rsa_key(user_private_key_pem, password)
+      user_ciphertext = encrypt_with_key(plaintext, user_private_key)
```

Comment:
the PII JSON string is encrypted using the one-time user-specific RSA private key. The JSON is encrypted with AES, and the random CEK generated for the AES encryption is encrypted with the RSA private key. See `encrypt_with_key` below.